### PR TITLE
Intermediate Type instantiation

### PR DESCRIPTION
**Pull request by mtrberzi**
The code from my previous pull request is now "live" in the sense that the Types and TypeDefinitions can now be instantiated to create instances of the corresponding Value objects. There are also a large number of test cases added to this code, bringing code coverage up to 75% (we are still missing a few rudimentary tests on FooTypes, mostly equality tests, but nothing major that is stopping others from at least starting to look over this PR).

### CODE REVIEW

File: src/main/java/org/whdl/intermediate/ArrayType.java [hunk 1]
```diff
@@ -42,4 +41,9 @@ public boolean equals(Object obj) {
     }
     return true;
   }
+
+  @Override
+  public Value instantiate(String instanceName) {
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/ArrayType.java]: What's the purpose of `instanceName`?

File: src/main/java/org/whdl/intermediate/ArrayValue.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+public class ArrayValue extends Value {
+
+  private Type elementType;
+  
+  public ArrayValue(Type elementType, String instanceName){
+    super(new ArrayType(elementType), instanceName);
+    this.elementType = elementType;
+  }
+  
+  // FIXME functions and internals on arrays: getting and setting their contents, getting length, etc.
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/ArrayValue.java]: For an initial API, lets support `void get(int)` and `int getLength()`

File: src/main/java/org/whdl/intermediate/ArrayValue.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package org.whdl.intermediate;
+
+public class ArrayValue extends Value {
+
+  private Type elementType;
+  
+  public ArrayValue(Type elementType, String instanceName){
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/ArrayValue.java]: I think we ought to pass all array values to the constructor. Making classes immutable leads to fewer bugs. And it makes sense given that our domain objects are all immutable.

File: src/main/java/org/whdl/intermediate/BooleanValue.java [hunk 4]
```diff
@@ -1,17 +1,14 @@
...
-import org.whdl.intermediate.types.PrimitiveType;
+import org.whdl.intermediate.PrimitiveType.PrimitiveKind;
+
 
 public class BooleanValue extends Value {
 
   private Boolean val;
-  public BooleanValue(Boolean val){
+  public BooleanValue(String instanceName, Boolean val){
+    super(new PrimitiveType(PrimitiveKind.BOOLEAN), instanceName);
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/BooleanValue.java]: Why can't we just have a `BooleanValue` class that references a `BooleanType` class which inherits from a `Type` class? That's more consistent with the front-end, simplifies this interface, and seems to be more amenable to extension (see [open/closed principle](http://en.wikipedia.org/wiki/Open/closed_principle))

File: src/main/java/org/whdl/intermediate/Connection.java [hunk 5]
```diff
@@ -4,25 +4,28 @@
...
+  }
+  public void setAttribute(String attrName, Value attrValue){
+    attributes.put(attrName, attrValue);
+  }
+  private Endpoint endpointFrom = null, endpointTo = null;
   
-  public Connection(String instanceName, ConnectionTypeDefinition definition, Endpoint endpointFrom, Endpoint endpointTo){
-    this.instanceName = instanceName;
+  public Connection(String instanceName, ConnectionTypeDefinition definition){
+    super(new ConnectionType(definition), instanceName);
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Connection.java]: Didn't we decide that we didn't need to declare `ConnectionType`s? All connections are of one type, if greater granularity is needed, `Nodes` are used.

File: src/main/java/org/whdl/intermediate/ArrayType.java (hunk 1 above)
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/ArrayType.java]: Lets put `instanceNames` in symbol tables

**Comment by lucaswoj** [comment]
- [x] create symbol table and remove the notion of `instanceName`
- [x] remove the notion of a `TypeDefinition` and make every value reference its type, every value of the same type references the same type object

File: src/main/java/org/whdl/intermediate/ArrayType.java (hunk 1 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/ArrayType.java]: Got it, next round of commits will have no `instanceName`s.

**Comment by mtrberzi** [comment]
The symbol table is not in scope for this pull request and will be dealt with in a future branch.

File: src/main/java/org/whdl/intermediate/ArrayValue.java (hunk 3 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/ArrayValue.java]: This would be great, but it changes the semantics of `Type.instantiate()`, because we expect to be able to call that with no arguments regardless of what we are instantiating.

**Comment by mtrberzi** [comment]
New commits are available! Please have a look at your leisure.

File: src/main/java/org/whdl/intermediate/ArrayType.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package org.whdl.intermediate;
+
+
+public class ArrayType extends Type {
+  private Type elementType;
+  public Type getElementType(){
+    return this.elementType;
+  }
+  
+  public ArrayType(String typename, Type elementType){
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/ArrayType.java]: We've gotten rid of `instancename` but not `typename`?

File: src/main/java/org/whdl/intermediate/ArrayValue.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ArrayValue extends Value {
+
+  private Type elementType;
+  private List<Value> contents = null;
+  
+  public ArrayValue(ArrayType t){
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/ArrayValue.java]: Are arrays immutable? If so, I think we should pass the values in the constructor.

File: src/main/java/org/whdl/intermediate/ArrayValue.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    super(t);
+    this.elementType = t.getElementType();
+  }
+  
+  public void setContents(List<Value> newContents) throws TypeException{
+	  // type-check contents -- every Value must have type 'elementType'
+	  for(Value v : newContents){
+		  Type t = v.getType();
+		  if(!t.equals(elementType)){
+			  throw new TypeException(elementType.toString(), t.toString());
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/ArrayValue.java]: Maybe we can share some exceptions between the frontend and backend? If not, perhaps you could look at the names we're using in the front-end and mirror them in the backend.

File: src/main/java/org/whdl/intermediate/BooleanValue.java [hunk 9]
```diff
@@ -1,17 +1,14 @@
 package org.whdl.intermediate;
 
-import org.whdl.intermediate.types.PrimitiveType;
+import org.whdl.intermediate.PrimitiveType.PrimitiveKind;
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/BooleanValue.java]: I think the whole system would be a little cleaner if we just bite the bullet and create `Type` subclasses for all of the `PrimitiveKind` values

File: src/main/java/org/whdl/intermediate/Constraint.java [hunk 10]
```diff
@@ -4,22 +4,22 @@
 import java.util.Map;
 
 public class Constraint extends Value {
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Constraint.java:6]: I'm still wondering if there's a common abstraction underlying `Constraint`s and `Annotation`s.

File: src/main/java/org/whdl/intermediate/ConstraintType.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+      String argName = arg.getKey();
+      UserDefinedType argTypeValue = arg.getValue();
+      Value argValue = argTypeValue.instantiate();
+      con.setArgument(argName, argValue);
+    }
+    return con;
+  }
+  
+  public void addArgument(String argName, UserDefinedType argType) {
+    // FIXME multiple additions of the same attribute?
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/ConstraintType.java]: Why not pass in the arguments to `instantiate`? That would solve this problem.

File: src/main/java/org/whdl/intermediate/EndpointType.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+      String attrName = attr.getKey();
+      UserDefinedType attrTypeValue = attr.getValue();
+      Value attrValue = attrTypeValue.instantiate();
+      ept.setAttribute(attrName, attrValue);
+    }
+    return ept;
+  }
+
+  public void addAttribute(String attrName, UserDefinedType attrType) {
+    // FIXME multiple additions of the same attribute?
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/EndpointType.java]: Same as above

File: src/main/java/org/whdl/intermediate/Node.java [hunk 13]
```diff
@@ -4,24 +4,34 @@
...
-  private Map<String, Value> attributes;
   private Map<String, Endpoint> endpoints;
+  public Endpoint getEndpoint(String eptName) throws UndeclaredIdentifierException{
+    if(endpoints.containsKey(eptName)){
+      return endpoints.get(eptName);
+    }else{
+      throw new UndeclaredIdentifierException("no endpoint named '" + eptName + "'");
+    }
+  }
+  public void setEndpoint(String eptName, Endpoint eptValue){
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Node.java:27]: Shouldn't `NodeType` define the endpoints?

File: src/main/java/org/whdl/intermediate/NodeType.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+      String attrName = attr.getKey();
+      UserDefinedType attrTypeValue = attr.getValue();
+      Value attrValue = attrTypeValue.instantiate();
+      node.setAttribute(attrName, attrValue);
+    }
+    return node;
+  }
+  
+  public void addAttribute(String attrName, UserDefinedType attrType) {
+    // FIXME multiple additions of the same attribute?
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/NodeType.java]: It seems like this pattern reoccurs very often. Perhaps we could create an `AttributedType` base class?

File: src/main/java/org/whdl/intermediate/PrimitiveType.java [hunk 15]
```diff
@@ -40,6 +40,21 @@ public boolean equals(Object obj) {
     }
     return true;
   }
+
+  @Override
+  public Value instantiate() {
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/PrimitiveType.java]: This doesn't seem particularly useful as all these values should be immutable after they are created

File: src/main/java/org/whdl/intermediate/Schematic.java [hunk 16]
```diff
@@ -30,13 +26,13 @@
...
     populateDefaultTypeDefinitions();
 
-    this.endpointTypeDefinitions = new HashMap<String, EndpointTypeDefinition>();
-    this.nodeTypeDefinitions = new HashMap<String, NodeTypeDefinition>();
-    this.connectionTypeDefinitions = new HashMap<String, ConnectionTypeDefinition>();
-    this.constraintTypeDefinitions = new HashMap<String, ConstraintTypeDefinition>();
+    this.endpointTypes = new HashMap<String, EndpointType>();
+    this.nodeTypes = new HashMap<String, NodeType>();
+    this.connectionTypes = new HashMap<String, ConnectionType>();
+    this.constraintTypes = new HashMap<String, ConstraintType>();
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Schematic.java:35]: I think these all should go into your symbol table class eventually.

File: src/main/java/org/whdl/intermediate/Schematic.java [hunk 17]
```diff
@@ -49,9 +45,9 @@ public Schematic(String name){
    * Every class in .intermediate.types should be represented in here.
    */
   private void populateDefaultTypeDefinitions(){
-    TypeTypeDefinition boolType = new TypeTypeDefinition("Bool", new PrimitiveType(PrimitiveType.PrimitiveKind.BOOLEAN));
-    TypeTypeDefinition intType = new TypeTypeDefinition("Int", new PrimitiveType(PrimitiveType.PrimitiveKind.INTEGER));    
-    TypeTypeDefinition stringType = new TypeTypeDefinition("String", new PrimitiveType(PrimitiveType.PrimitiveKind.STRING));
+    UserDefinedType boolType = new UserDefinedType("Bool", new PrimitiveType(PrimitiveType.PrimitiveKind.BOOLEAN));
+    UserDefinedType intType = new UserDefinedType("Int", new PrimitiveType(PrimitiveType.PrimitiveKind.INTEGER));    
+    UserDefinedType stringType = new UserDefinedType("String", new PrimitiveType(PrimitiveType.PrimitiveKind.STRING));
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Schematic.java]: It should be possible to put a `PrimitiveType` in the symbol table without wrapping it in a `UserDefinedType`. `UserDefinedType`s should be reserved for actual types created by the user.

File: src/main/java/org/whdl/intermediate/Schematic.java [hunk 18]
```diff
@@ -62,44 +58,84 @@ private void populateDefaultTypeDefinitions(){
     }
   }
   
-  public void addTypeTypeDefinition(TypeTypeDefinition td) throws MultipleDefinitionException{
+  public void addTypeTypeDefinition(UserDefinedType td) throws MultipleDefinitionException{
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Schematic.java]: I think you intended to rename this method. (Although this comment is moot if you move `Type`s to the symbol table)

File: src/main/java/org/whdl/intermediate/Schematic.java [hunk 19]
```diff
@@ -62,44 +58,84 @@ private void populateDefaultTypeDefinitions(){
...
+    constraintTypes.put(key, cd);
+  }
+  
+  public ConstraintType getConstraintTypeDefinition(String typename) throws UndeclaredIdentifierException {
+    if(constraintTypes.containsKey(typename)){
+      return constraintTypes.get(typename);
+    }else{
+      throw new UndeclaredIdentifierException(typename);
+    }
   }
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Schematic.java:134]: A lot of this code can disappear and be rewritten more generally inside our symbol table :smile:

File: src/main/java/org/whdl/intermediate/TypeException.java [hunk 20]
```diff
@@ -0,0 +1,17 @@
+package org.whdl.intermediate;
+
+public class TypeException extends Exception {
+	private static final long serialVersionUID = 7730618233489002412L;
+	private String expected;
+	private String actual;
+	
+	public TypeException(String expectedTypename, String actualTypename){
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/TypeException.java]: Might it be useful to pass in the `Type` itself for further introspection?

**Comment by lucaswoj** [comment]
We're getting really close on this one! Most of my comments fit into these categories
- moving `typename`s to the symbol table the way we did `instanceName``
- getting rid of `PrimitiveKind`
- ideas for future projects (anything that looks too big for the next round of revisions we'll open a new issue for)
- language name ideas

Thanks for all your hard work Murphy! I hope to have something concrete to show at our team meeting this week in terms of a language spec and roadmap for this semester.

File: src/main/java/org/whdl/intermediate/Connection.java [hunk 21]
```diff
@@ -4,25 +4,23 @@
...
-  private Endpoint endpointFrom, endpointTo;
+  public Value getAttribute(String attrName) throws UndeclaredIdentifierException{
+    if(attributes.containsKey(attrName)){
+      return attributes.get(attrName);
+    }else{
+      throw new UndeclaredIdentifierException("no attribute named '" + attrName + "'");
+    }
+  }
+  public void setAttribute(String attrName, Value attrValue){
+    attributes.put(attrName, attrValue);
```
Comment (TysonAndre) [inline src/main/java/org/whdl/intermediate/Connection.java:17]: Should setAttribute be limited to the attributes the ConnectionType was declared with?
Otherwise, you could have different Connections of the same ConnectionType with a string/boolean/integer for a given attrName.

File: src/test/java/org/whdl/intermediate/TestConnection.java [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+    ept.setAttribute("v", v);
+  }
+  
+  @Test
+  public void testSetAttribute_multiple_set() {
+    // setting an attribute twice should just work
+    Connection ept = new Connection(defaultConnectionDefinition);
+    Value v = new BooleanValue(true);
+    ept.setAttribute("v", v);
+    Value v2 = new BooleanValue(false);
```
Comment (TysonAndre) [inline src/test/java/org/whdl/intermediate/TestConnection.java]: Test setting as a StringValue?

Comment (lucaswoj) [inline src/test/java/org/whdl/intermediate/TestConnection.java]: Also, what is the use case for changing the value of an attribute? This is not supported in the front end language.

File: src/main/java/org/whdl/intermediate/ArrayType.java (hunk 6 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/ArrayType.java]: Having thought about this...yeah, it's unnecessary. Removing.

File: src/main/java/org/whdl/intermediate/ArrayValue.java (hunk 8 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/ArrayValue.java]: We can't mix them because they will deal with different objects. TypeException has become TypeMismatchException, and it now takes two types (expected and actual) instead of their typenames, which is much better for debugging. However, the front-end exception will never deal with intermediate types, and so we need a separate class.

File: src/main/java/org/whdl/intermediate/BooleanValue.java (hunk 9 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/BooleanValue.java]: Fair enough. This is done. On a related note, I'm seeing a flaw in the interface for `Type`, more on this later.

File: src/main/java/org/whdl/intermediate/Constraint.java (hunk 10 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/Constraint.java:6]: What do you mean by `Annotation`? We don't have that in the intermediate.

File: src/main/java/org/whdl/intermediate/ConstraintType.java (hunk 11 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/ConstraintType.java]: Right, and that's what I mentioned earlier about seeing a flaw in `Type`'s interface, in which `instantiate()` always takes no arguments. I'm removing that. Now, instantiating a type requires knowing what value it will have, since intermediate values are immutable. The problem with this is user-defined types, which make things more complicated because we don't know what arguments to pass to the "subtype" any more. So the actual solution is to remove `UserDefinedType` altogether, because it does nothing anymore now that we have removed type definitions.

**Comment by mtrberzi** [comment]
As I mentioned earlier, the symbol table is out of scope for this PR and will come up in the future. I know there are still a few outstanding comments I haven't acknowledged yet, but I just did a LOT of refactoring and I'd like to get some feedback on what I think is a cleaner design. `Type.instantiate()` and `UserDefinedType` are both GONE, along with `PrimitiveType`. New commits will be available very shortly, just as soon as I fix all the test cases again.

File: src/main/java/org/whdl/intermediate/TypeException.java (hunk 20 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/TypeException.java]: Yup, got this now.

File: build.gradle [hunk 23]
```diff
@@ -16,7 +16,8 @@ buildscript {
 apply plugin: 'jacoco'
 
 jacoco {
-  toolVersion = "0.6.3.201306030806"
+//  toolVersion = "0.6.3.201306030806"
+  toolVersion = "0.7.1.201405082137"
```
Comment (lucaswoj) [inline build.gradle]: Why did this happen?

File: src/main/java/org/whdl/intermediate/BooleanValue.java [hunk 24]
```diff
@@ -1,17 +1,11 @@
 package org.whdl.intermediate;
 
-import org.whdl.intermediate.types.PrimitiveType;
-
 public class BooleanValue extends Value {
 
   private Boolean val;
-  public BooleanValue(Boolean val){
+  public BooleanValue(Type t, Boolean val){
+    super(t);
```
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/BooleanValue.java:7]: Why can't we just pass `BooleanType` to `super` statically?

File: src/main/java/org/whdl/intermediate/Connection.java (hunk 21 above)
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Connection.java:17]: Tyson has a good point. Valid attribute identifiers should be statically declared in the `ConnectionType` instance and enforced all the time.

File: src/main/java/org/whdl/intermediate/Constraint.java (hunk 10 above)
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Constraint.java:6]: Sorry. I guess by `Annotation` I mean "`attribute`". There are lot of classes that have some collection of statically declared key/value pairs. Seems like a good candidate for abstraction. We can do that in a separate commit, if you agree its a good idea.

File: src/main/java/org/whdl/intermediate/Node.java (hunk 13 above)
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Node.java:27]: _bump_

**Comment by lucaswoj** [comment]
So close. I think I'd be happy with textual answers to my questions, opening issues for outstanding tasks, and merging this today. :ship: it

File: build.gradle (hunk 23 above)
Comment (mtrberzi) [inline build.gradle]: Gradle somehow broke for me when using that version -- there was a bizarre reflection error and nothing I could do short of bumping the version fixed the error.

File: src/main/java/org/whdl/intermediate/Connection.java (hunk 21 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/Connection.java:17]: This creates #80.

File: src/main/java/org/whdl/intermediate/Schematic.java (hunk 16 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/Schematic.java:35]: This creates #79.

File: src/main/java/org/whdl/intermediate/Node.java (hunk 13 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/Node.java:27]: Can you elaborate on that a bit? I don't quite follow that thought in this context.

File: src/main/java/org/whdl/intermediate/Constraint.java (hunk 10 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/Constraint.java:6]: This creates #81.

File: src/main/java/org/whdl/intermediate/BooleanValue.java (hunk 24 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/BooleanValue.java:7]: `BooleanType` isn't a singleton object -- should it be? If so, we can pass it statically. If not, we need the instance of `BooleanType` the schematic/symbol table gives us. Will defer on opening an issue until we have a decision (I would prefer the former if it is correct).

Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/BooleanValue.java:7]: `BooleanType` should probably be a singleton object. You can add a static `getInstance()` method, as we do for front-end classes.

File: src/main/java/org/whdl/intermediate/Node.java (hunk 13 above)
Comment (lucaswoj) [inline src/main/java/org/whdl/intermediate/Node.java:27]: I think I misunderstood what was happening here. You should probably check these values against the defined endpoint names but the behavior is good.

File: build.gradle (hunk 23 above)
Comment (lucaswoj) [inline build.gradle]: Ok. Could you please revert this and submit a separate pull request for this change.

File: src/main/java/org/whdl/intermediate/BooleanValue.java (hunk 24 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/BooleanValue.java:7]: Okay, this is now #82.
(And I really need a new issue label for IR.)

File: src/main/java/org/whdl/intermediate/Node.java (hunk 13 above)
Comment (mtrberzi) [inline src/main/java/org/whdl/intermediate/Node.java:27]: Right, that counts as attribute "type-checking", which is already #80.

File: build.gradle (hunk 23 above)
Comment (mtrberzi) [inline build.gradle]: Coming right up...

**Comment by mtrberzi** [comment]
I think this is ready to go (after the revert), any last requests?

**Comment by lucaswoj** [comment]
Nope. Merge away.

**Comment by lucaswoj** [comment]
:fireworks:
